core/src/broadcast_stage/broadcast_utils.rs

```diff
@@ -26,45 +30,63 @@ pub struct UnfinishedSlotInfo {
     pub parent: Slot,
 }
 
-/// This parameter tunes how many entries are received in one iteration of recv loop
-/// This will prevent broadcast stage from consuming more entries, that could have led
-/// to delays in shredding, and broadcasting shreds to peer validators
-const RECEIVE_ENTRY_COUNT_THRESHOLD: usize = 8;
-
 pub(super) fn recv_slot_entries(receiver: &Receiver<WorkingBankEntry>) -> Result<ReceiveResults> {
+    let target_serialized_batch_byte_count: u64 =
+        32 * ShredData::capacity(/*merkle_proof_size*/ None).unwrap() as u64;
+    let mut serialized_batch_byte_count: u64 = 8; // Vec len
     let timer = Duration::new(1, 0);
     let recv_start = Instant::now();
+
     let (mut bank, (entry, mut last_tick_height)) = receiver.recv_timeout(timer)?;
 
+    serialized_batch_byte_count += serialized_size(&entry)?;
     let mut entries = vec![entry];
-    let mut slot = bank.slot();
-    let mut max_tick_height = bank.max_tick_height();
-
-    assert!(last_tick_height <= max_tick_height);
-
-    if last_tick_height != max_tick_height {
-        while let Ok((try_bank, (entry, tick_height))) = receiver.try_recv() {
-            // If the bank changed, that implies the previous slot was interrupted and we do not have to
-            // broadcast its entries.
-            if try_bank.slot() != slot {
-                warn!("Broadcast for slot: {} interrupted", bank.slot());
-                entries.clear();
-                bank = try_bank;
-                slot = bank.slot();
-                max_tick_height = bank.max_tick_height();
-            }
-            last_tick_height = tick_height;
-            entries.push(entry);
-
-            if entries.len() >= RECEIVE_ENTRY_COUNT_THRESHOLD {
-                break;
-            }
-
-            assert!(last_tick_height <= max_tick_height);
-            if last_tick_height == max_tick_height {
-                break;
-            }
+
+    assert!(last_tick_height <= bank.max_tick_height());
+
+    // Drain channel
+    while last_tick_height != bank.max_tick_height() {
+        let (try_bank, (entry, tick_height)) = match receiver.try_recv() {
+            Ok(working_bank_entry) => working_bank_entry,
+            Err(_) => break,
+        };
+        // If the bank changed, that implies the previous slot was interrupted and we do not have to
+        // broadcast its entries.
+        if try_bank.slot() != bank.slot() {
+            warn!("Broadcast for slot: {} interrupted", bank.slot());
+            entries.clear();
+            serialized_batch_byte_count = 8; // Vec len
+            bank = try_bank;
+        }
+        last_tick_height = tick_height;
+        let entry_bytes = serialized_size(&entry)?;
+        serialized_batch_byte_count += entry_bytes;
+        entries.push(entry);
+        assert!(last_tick_height <= bank.max_tick_height());
+    }
```

Comment:
No need to track `serialized_batch_byte_count` in this loop.
Can just do
```rust
let mut serialized_batch_byte_count = bincode::serialized_size(&entries);
```
after the loop which would be more succinct.